compiler/src/iree/compiler/GlobalOptimization/FuseHorizontalContractions.cpp

```diff
@@ -0,0 +1,424 @@
+    Value truncOuts = rewriter.create<tensor::EmptyOp>(
+        loc, concatTruncType, origEmpty.getDynamicSizes());
+    SmallVector<AffineMap> truncIndexingMaps(
+        2, AffineMap::getMultiDimIdentityMap(concatTruncType.getRank(),
+                                             rewriter.getContext()));
+    SmallVector<utils::IteratorType> truncIteratorTypes(
+        concatTruncType.getRank(), utils::IteratorType::parallel);
+    auto truncateOpBody = [&](OpBuilder &b, Location loc, ValueRange args) {
+      Value truncOp = b.create<arith::TruncFOp>(
+          loc, concatTruncType.getElementType(), args[0]);
+      rewriter.create<linalg::YieldOp>(loc, truncOp);
+    };
+    auto truncateOp = rewriter.create<linalg::GenericOp>(
+        loc, concatTruncType, newGenericOp->getResults(), truncOuts,
+        truncIndexingMaps, truncIteratorTypes, truncateOpBody);
+
+    fusedResult = truncateOp.getResult(0);
+  }
+
+  SmallVector<Value> newOuts;
+
+  auto fusedResultType = cast<RankedTensorType>(fusedResult.getType());
+  SmallVector<int64_t> shape = llvm::to_vector(fusedResultType.getShape());
+
+  SmallVector<OpFoldResult> sizes = newEmpty.getMixedSizes();
+  sizes[0] = rewriter.getIndexAttr(1);
+
+  SmallVector<OpFoldResult> offsets(fusedResultType.getRank(),
+                                    rewriter.getIndexAttr(0));
+  SmallVector<OpFoldResult> strides(fusedResultType.getRank(),
+                                    rewriter.getIndexAttr(1));
+  auto resultOutType = RankedTensorType::get(outType.getShape(),
+                                             fusedResultType.getElementType());
+  for (auto i : llvm::seq<int>(0, rhsVals.size())) {
+    offsets[0] = rewriter.getIndexAttr(i);
+    newOuts.push_back(rewriter.create<tensor::ExtractSliceOp>(
+        loc, resultOutType, fusedResult, offsets, sizes, strides));
+  }
+
+  for (auto [index, op, replacement] :
+       llvm::enumerate(fusionGroup.contractionOps, newOuts)) {
+    Operation *replacedOp = op;
+    if (fusionGroup.truncateOps) {
+      replacedOp = fusionGroup.truncateOps.value()[index];
+    }
+    rewriter.replaceOp(replacedOp, replacement);
+  }
+  return success();
+}
+
+void FuseHorizontalContractionsPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  DominanceInfo dominanceInfo(getOperation());
+
+  SmallVector<HorizontalFusionGroup> horizontalFusionGroups;
+  llvm::SmallDenseSet<linalg::LinalgOp> groupedOperations;
+
+  getOperation()->walk([&](linalg::LinalgOp linalgOp) {
+    if (!linalg::isaContractionOpInterface(linalgOp)) {
+      return;
+    }
+    // Avoid already grouped operations;
+    if (groupedOperations.contains(linalgOp)) {
+      return;
+    }
+
+    std::optional<linalg::GenericOp> truncUser = getTruncUser(linalgOp);
+
+    std::optional<HorizontalFusionGroup> fusionGroup =
+        getHorizontalFusionGroupMembers(linalgOp, truncUser, groupedOperations,
+                                        dominanceInfo);
```

Comment:
I think we don't need `truncUser` here. It is only used by `getHorizontalFusionGroupMembers`. Instead, can we get the trunc user in the `getHorizontalFusionGroupMembers` method?

Reply:
Good point.